compiler/rustc_builtin_macros/src/concat_bytes.rs

```diff
@@ -0,0 +1,210 @@
+use rustc_ast as ast;
+use rustc_ast::tokenstream::TokenStream;
+use rustc_data_structures::sync::Lrc;
+use rustc_errors::Applicability;
+use rustc_expand::base::{self, DummyResult};
+
+pub fn expand_concat_bytes(
+    cx: &mut base::ExtCtxt<'_>,
+    sp: rustc_span::Span,
+    tts: TokenStream,
+) -> Box<dyn base::MacResult + 'static> {
+    let es = match base::get_exprs_from_tts(cx, sp, tts) {
+        Some(e) => e,
+        None => return DummyResult::any(sp),
+    };
+    let mut accumulator = Vec::new();
+    let mut missing_literal = vec![];
+    let mut has_errors = false;
+    for e in es {
+        match e.kind {
+            ast::ExprKind::Array(ref exprs) => {
+                for expr in exprs {
+                    match expr.kind {
+                        ast::ExprKind::Array(_) => {
+                            cx.span_err(expr.span, "cannot concatenate doubly nested array");
+                            has_errors = true;
```

Comment:
Hm, can we bundle up the spans for the wrong-type errors and give a single multi-span error or something? Otherwise it gets a bit noisy.

Alternatively it might make sense to stop complaining about further errors after the first one...

Reply:
I made it only emit one invalid literal error. A multispan with all of the invalid literals would still be pretty noisy due to the suggestions getting their own sections.